File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/widget/ScheduleWidget.java
package com.vereshchagin.nikolay.stankinschedule.widget;

import android.app.PendingIntent;
import android.appwidget.AppWidgetManager;
import android.appwidget.AppWidgetProvider;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Bundle;
import android.widget.RemoteViews;

import androidx.navigation.NavDeepLinkBuilder;

import com.vereshchagin.nikolay.stankinschedule.MainActivity;
import com.vereshchagin.nikolay.stankinschedule.R;
import com.vereshchagin.nikolay.stankinschedule.schedule.view.ScheduleViewFragment;
import com.vereshchagin.nikolay.stankinschedule.settings.SchedulePreference;

import java.util.Calendar;

/**
 * Виджет с расписанием.
 */
public class ScheduleWidget extends AppWidgetProvider {

    public static final String SCHEDULE_DAY_TIME = "widget_schedule_day_time";
    public static final String SCHEDULE_NAME = "widget_schedule_name";

    private static final String ACTION_SCHEDULE_DAY_CLICKED = "action_schedule_day_clicked";

    /**
     * Обновляет данные на виджете с расписанием.
     * @param context контекст.
     * @param appWidgetManager менеджер виджетов.
     * @param appWidgetId ID виджета.
     */
    public static void updateAppWidget(Context context, AppWidgetManager appWidgetManager,
                                int appWidgetId) {

        ScheduleWidgetConfigureActivity.WidgetData widgetData =
                ScheduleWidgetConfigureActivity.loadPref(context, appWidgetId);

        String scheduleName = widgetData.scheduleName();
        if (scheduleName == null) {
            scheduleName = context.getString(R.string.widget_schedule_name);
        }

        RemoteViews views = new RemoteViews(context.getPackageName(), R.layout.widget_schedule);

        // для открытия приложения на распиании
        Bundle scheduleBundle = ScheduleViewFragment.createBundle(scheduleName,
                SchedulePreference.createPath(context, scheduleName));

        PendingIntent schedulePendingIntent = new NavDeepLinkBuilder(context)
                .setComponentName(MainActivity.class)
                .setGraph(R.navigation.activity_main_nav_graph)
                .setDestination(R.id.nav_schedule_view_fragment)
                .setArguments(scheduleBundle)
                .createPendingIntent();

        views.setTextViewText(R.id.widget_schedule_name, scheduleName);
        views.setOnClickPendingIntent(R.id.widget_schedule_name, schedulePendingIntent);

        // для открытия приложения на расписании на определенном дне
        Intent scheduleDayIntent = new Intent(context, ScheduleWidget.class);
        scheduleDayIntent.setAction(ACTION_SCHEDULE_DAY_CLICKED);
        PendingIntent scheduleDayPendingIntent = PendingIntent
                .getBroadcast(context, appWidgetId, scheduleDayIntent, 0);

        views.setPendingIntentTemplate(R.id.widget_schedule_list, scheduleDayPendingIntent);

        // установка адаптера
        Intent dataIntent = new Intent(context, ScheduleWidgetRemoteFactory.Service.class);
        dataIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId);
        // уникальный адаптер для каждого виджета
        Uri data = Uri.parse(dataIntent.toUri(Intent.URI_INTENT_SCHEME));
        dataIntent.setData(data);

        views.setRemoteAdapter(R.id.widget_schedule_list, dataIntent);

        // Обновление виджета
        appWidgetManager.updateAppWidget(appWidgetId, views);
        appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetId, R.id.widget_schedule_list);
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        super.onReceive(context, intent);

        String action = intent.getAction();
        if (action != null && action.equalsIgnoreCase(ACTION_SCHEDULE_DAY_CLICKED)) {

            Calendar date = (Calendar) intent.getSerializableExtra(SCHEDULE_DAY_TIME);
            String scheduleName =  intent.getStringExtra(SCHEDULE_NAME);

            if (scheduleName == null) {
                return;
            }

            // создание intent'а на открытие расписание на определенном дне
            Bundle scheduleDayBundle = ScheduleViewFragment.createBundle(scheduleName,
                    SchedulePreference.createPath(context, scheduleName), date);

            PendingIntent scheduleDayPendingIntent = new NavDeepLinkBuilder(context)
                    .setComponentName(MainActivity.class)
                    .setGraph(R.navigation.activity_main_nav_graph)
                    .setDestination(R.id.nav_schedule_view_fragment)
                    .setArguments(scheduleDayBundle)
                    .createPendingIntent();

            try {
                scheduleDayPendingIntent.send();
            } catch (PendingIntent.CanceledException ignored) {

            }
        }
    }

    @Override
    public void onUpdate(Context context, AppWidgetManager appWidgetManager, int[] appWidgetIds) {
        // Обновить все виджеты
        for (int appWidgetId : appWidgetIds) {
            updateAppWidget(context, appWidgetManager, appWidgetId);
        }
    }

    @Override
    public void onDeleted(Context context, int[] appWidgetIds) {
        // Удалить настройки , связанные с этими виджетами
        for (int appWidgetId : appWidgetIds) {
            ScheduleWidgetConfigureActivity.deletePref(context, appWidgetId);
        }
    }

    @Override
    public void onEnabled(Context context) {
        // Вызывается когда установлен первый виджет
    }

    @Override
    public void onDisabled(Context context) {
        // Вызывается когда удален последний виджет
        // ApplicationPreference.clearScheduleWidgetIDs(context);
    }
}

File: app/src/main/java/com/vereshchagin/nikolay/stankinschedule/widget/ScheduleWidget.kt
package com.vereshchagin.nikolay.stankinschedule.widget

import android.app.PendingIntent
import android.app.PendingIntent.CanceledException
import android.appwidget.AppWidgetManager
import android.appwidget.AppWidgetProvider
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.util.Log
import android.widget.RemoteViews
import androidx.navigation.NavDeepLinkBuilder
import com.google.firebase.crashlytics.FirebaseCrashlytics
import com.vereshchagin.nikolay.stankinschedule.MainActivity
import com.vereshchagin.nikolay.stankinschedule.R
import com.vereshchagin.nikolay.stankinschedule.repository.ScheduleRepository
import com.vereshchagin.nikolay.stankinschedule.ui.schedule.view.ScheduleViewFragment
import com.vereshchagin.nikolay.stankinschedule.utils.extensions.FLAG_MUTABLE_COMPAT
import com.vereshchagin.nikolay.stankinschedule.widget.paging.ScheduleWidgetListService
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.withContext
import org.joda.time.LocalDate

/**
 * Виджет с расписанием.
 */
class ScheduleWidget : AppWidgetProvider() {

    override fun onReceive(context: Context, intent: Intent) {
        super.onReceive(context, intent)

        val action = intent.action
        if (action != null && action.equals(ACTION_SCHEDULE_DAY_CLICKED, ignoreCase = true)) {

            Log.d(TAG, "onReceive: $intent")
            Log.d(TAG, "onReceive: ${intent.extras}")

            val date = intent.getSerializableExtra(SCHEDULE_DAY_TIME) as LocalDate? ?: return
            val scheduleId = intent.getLongExtra(SCHEDULE_ID, -1)

            // создание intent'а на открытие расписание на определенном дне
            val scheduleDayBundle = ScheduleViewFragment.createBundle(scheduleId, date)

            val scheduleDayPendingIntent = NavDeepLinkBuilder(context)
                .setComponentName(MainActivity::class.java)
                .setGraph(R.navigation.activity_main_nav_graph)
                .setDestination(R.id.nav_schedule_view_fragment)
                .setArguments(scheduleDayBundle)
                .createTaskStackBuilder()
                .getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT or FLAG_MUTABLE_COMPAT)

            try {
                scheduleDayPendingIntent?.send()

            } catch (ignored: CanceledException) {

            }
        }
    }

    override fun onUpdate(
        context: Context,
        appWidgetManager: AppWidgetManager,
        appWidgetIds: IntArray,
    ) {
        try {
            // Обновить все виджеты
            for (appWidgetId in appWidgetIds) {
                updateScheduleWidget(context, appWidgetManager, appWidgetId)
            }
        } catch (t: Throwable) {
            FirebaseCrashlytics.getInstance().recordException(t)
            throw t
        }
    }

    override fun onDeleted(context: Context, appWidgetIds: IntArray) {
        // Удалить настройки , связанные с этими виджетами
        for (appWidgetId in appWidgetIds) {
            ScheduleWidgetConfigureActivity.deletePref(context, appWidgetId)
        }
    }

    override fun onEnabled(context: Context) {
        // Вызывается когда установлен первый виджет
    }

    override fun onDisabled(context: Context) {
        // Вызывается когда удален последний виджет
        // ApplicationPreference.clearScheduleWidgetIDs(context);
    }

    companion object {

        private const val SCHEDULE_DAY_TIME = "widget_schedule_day_time"
        private const val SCHEDULE_ID = "widget_schedule_id"
        private const val ACTION_SCHEDULE_DAY_CLICKED = "action_schedule_day_clicked"

        private const val TAG = "ScheduleWidgetLog"

        /**
         * Обновляет данные на виджете с расписанием.
         * @param context контекст.
         * @param appWidgetManager менеджер виджетов.
         * @param appWidgetId ID виджета.
         */
        fun updateScheduleWidget(
            context: Context,
            appWidgetManager: AppWidgetManager,
            appWidgetId: Int,
        ) {
            val widgetData = ScheduleWidgetConfigureActivity.loadPref(context, appWidgetId)
            updateScheduleWidgetData(context, appWidgetManager, appWidgetId, widgetData)
        }

        /**
         * Обновляет данные на виджете с помощью репозитория расписаний.
         * @param context контекст.
         * @param appWidgetManager менеджер виджетов.
         * @param appWidgetId ID виджета.
         * @param scheduleId ID расписания.
         * @param repository репозиторий с расписаниями.
         */
        suspend fun updateScheduleWidget(
            context: Context,
            appWidgetManager: AppWidgetManager,
            appWidgetId: Int,
            scheduleId: Long,
            repository: ScheduleRepository,
        ) {
            // получение новых данных
            val scheduleItem = withContext(Dispatchers.IO) {
                repository.scheduleItem(scheduleId).first()
            } ?: return

            // обновление настроек
            var widgetData = ScheduleWidgetConfigureActivity.loadPref(context, appWidgetId)
            widgetData = ScheduleWidgetData(
                scheduleItem.scheduleName,
                widgetData.scheduleId,
                widgetData.subgroup,
                widgetData.display
            )
            ScheduleWidgetConfigureActivity.savePref(context, appWidgetId, widgetData)

            updateScheduleWidgetData(context, appWidgetManager, appWidgetId, widgetData)
        }

        /**
         * Обновляет только список в виджете с расписанием
         * @param appWidgetManager менеджер виджетов.
         * @param appWidgetId ID виджета.
         */
        fun updateScheduleWidgetList(appWidgetManager: AppWidgetManager, appWidgetId: Int) {
            appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetId, R.id.widget_schedule_list)
        }

        /**
         * Обновляет данные в виджете с расписанием.
         * @param context контекст.
         * @param appWidgetManager менеджер виджетов.
         * @param appWidgetId ID виджета.
         * @param widgetData новые данные.
         */
        private fun updateScheduleWidgetData(
            context: Context,
            appWidgetManager: AppWidgetManager,
            appWidgetId: Int,
            widgetData: ScheduleWidgetData,
        ) {
            // для открытия приложения на расписании
            val schedulePendingIntent = NavDeepLinkBuilder(context)
                .setComponentName(MainActivity::class.java)
                .setGraph(R.navigation.activity_main_nav_graph)
                .setDestination(R.id.nav_schedule_view_fragment)
                .setArguments(
                    ScheduleViewFragment.createBundle(widgetData.scheduleId)
                )
                .createTaskStackBuilder()
                .getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT or FLAG_MUTABLE_COMPAT)

            // установка имени
            val scheduleName = widgetData.displayName(context)
            val views = RemoteViews(context.packageName, R.layout.widget_schedule)
            views.setTextViewText(R.id.widget_schedule_name, scheduleName)
            views.setOnClickPendingIntent(R.id.widget_schedule_name, schedulePendingIntent)

            // для открытия приложения на расписании на определенном дне
            val scheduleDayPendingIntent = PendingIntent.getBroadcast(
                context,
                appWidgetId,
                Intent(context, ScheduleWidget::class.java).also {
                    it.action = ACTION_SCHEDULE_DAY_CLICKED
                },
                PendingIntent.FLAG_UPDATE_CURRENT or FLAG_MUTABLE_COMPAT
            )
            views.setPendingIntentTemplate(R.id.widget_schedule_list, scheduleDayPendingIntent)

            // установка адаптера
            val dataIntent = Intent(context, ScheduleWidgetListService::class.java)
            dataIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_ID, appWidgetId)

            // уникальный адаптер для каждого виджета
            val data = Uri.parse(dataIntent.toUri(Intent.URI_INTENT_SCHEME))
            dataIntent.data = data
            views.setRemoteAdapter(R.id.widget_schedule_list, dataIntent)

            // Обновление виджета
            appWidgetManager.updateAppWidget(appWidgetId, views)
            appWidgetManager.notifyAppWidgetViewDataChanged(appWidgetId, R.id.widget_schedule_list)
        }

        /**
         * Создает intent для обратного вызова приложения с расписанием на определенном дне.
         */
        fun createDayIntent(scheduleId: Long, date: LocalDate) = Intent().apply {
            putExtra(SCHEDULE_ID, scheduleId)
            putExtra(SCHEDULE_DAY_TIME, date)
        }
    }
}